File: apps/companion/modules/fairth-background-upload/src/FairthBackgroundUploadModule.ts
import { NativeModule, requireNativeModule } from "expo";
import type { BackgroundUploadStatus } from "./FairthBackgroundUpload.types";

declare class FairthBackgroundUploadModule extends NativeModule<{}> {
  configure(configurationJson: string, token: string): Promise<void>;
  runNow(): Promise<void>;
  enqueueManualAssets(assetsJson: string): Promise<number>;
  enqueueSharedImages(filesJson: string): Promise<number>;
  getStatus(): Promise<BackgroundUploadStatus>;
  checkConnection(): Promise<boolean>;
}

export default requireNativeModule<FairthBackgroundUploadModule>("FairthBackgroundUpload");

File: companion/modules/fairth-background-upload/android/src/main/java/dev/fairth/backgroundupload/FairthBackgroundUploadModule.kt
package dev.fairth.backgroundupload

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class FairthBackgroundUploadModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("FairthBackgroundUpload")

    AsyncFunction("configure") { configurationJson: String, token: String ->
      val context = requireNotNull(appContext.reactContext) { "Android application context is unavailable." }
      val configuration = UploadConfiguration.parse(configurationJson)
      UploadConfiguration.save(context, configuration)
      SecureTokenStore(context).save(token)
      UploadDatabase(context).use { it.retryAuthenticationFailures() }
      UploadScheduler.configure(context, configuration)
    }

    AsyncFunction("runNow") {
      val context = requireNotNull(appContext.reactContext) { "Android application context is unavailable." }
      val configuration = UploadConfiguration.load(context) ?: error("Save upload settings before starting sync.")
      UploadScheduler.runNow(context, configuration)
    }

    AsyncFunction("enqueueManualAssets") { assetsJson: String ->
      val context = requireNotNull(appContext.reactContext) { "Android application context is unavailable." }
      UploadDatabase(context).use { database -> MediaScanner(context, database).enqueueManual(assetsJson) }
    }

    AsyncFunction("getStatus") {
      val context = requireNotNull(appContext.reactContext) { "Android application context is unavailable." }
      UploadDatabase(context).use { it.counts() }
    }
  }
}
